Sources/CoreMetrics/SystemMetrics.swift

```diff
@@ -0,0 +1,322 @@
+        internal let underlying: MetricsFactory
+
+        init(factory: MetricsFactory, options: SystemMetrics.Options) {
+            self.underlying = factory
+            self.timeInterval = options.interval
+            if let systemMetricsType = options.metricsType {
+                self.systemMetricsType = systemMetricsType
+            } else {
+                #if os(Linux)
+                self.systemMetricsType = SystemMetrics.LinuxProvider.self
+                #else
+                self.systemMetricsType = SystemMetrics.NOOPProvider.self
+                #endif
+            }
+            self.labels = options.labels
+
+            self.task = DispatchWorkItem(qos: .background, block: {
+                guard let metrics = self.systemMetricsType.readSystemMetrics() else { return }
+                Gauge(label: self.labels.label(for: \.virtualMemoryBytes)).record(metrics.virtualMemoryBytes)
+                Gauge(label: self.labels.label(for: \.residentMemoryBytes)).record(metrics.residentMemoryBytes)
+                Gauge(label: self.labels.label(for: \.startTimeSeconds)).record(metrics.startTimeSeconds)
+                Gauge(label: self.labels.label(for: \.cpuSecondsTotal)).record(metrics.cpuSeconds)
+                Gauge(label: self.labels.label(for: \.maxFileDescriptors)).record(metrics.maxFileDescriptors)
+                Gauge(label: self.labels.label(for: \.openFileDescriptors)).record(metrics.openFileDescriptors)
+            })
+        }
+
+        internal func cancelSystemMetrics() {
+            self.task?.cancel()
+            self.task = nil
+        }
+
+        internal func poll() {
+            self.queue.asyncAfter(deadline: .now() + self.timeInterval) {
+                guard let task = self.task else { return }
+                task.perform()
+                self.poll()
+            }
+        }
+        
+        func makeCounter(label: String, dimensions: [(String, String)]) -> CounterHandler {
+            self.underlying.makeCounter(label: label, dimensions: dimensions)
+        }
+        
+        func makeRecorder(label: String, dimensions: [(String, String)], aggregate: Bool) -> RecorderHandler {
+            self.underlying.makeRecorder(label: label, dimensions: dimensions, aggregate: aggregate)
+        }
+        
+        func makeTimer(label: String, dimensions: [(String, String)]) -> TimerHandler {
+            self.underlying.makeTimer(label: label, dimensions: dimensions)
+        }
+        
+        func destroyCounter(_ handler: CounterHandler) {
+            self.underlying.destroyCounter(handler)
+        }
+        
+        func destroyRecorder(_ handler: RecorderHandler) {
+            self.underlying.destroyRecorder(handler)
+        }
+        
+        func destroyTimer(_ handler: TimerHandler) {
+            self.underlying.destroyTimer(handler)
+        }
+    }
+}
+
+public protocol SystemMetricsProvider {
+    static func readSystemMetrics() -> SystemMetrics.Data?
```

Comment:
is there a strong reason to use static func? I'd even argue that this can be simplified to:

`public typealias SystemMetrics.DataProvider = () -> SystemMetrics.Data?`, and then the user can provide a generator of their choice